compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp

```diff
@@ -0,0 +1,424 @@
+
+  linalg::LinalgOp base = fusionGroup.contractionOps.front();
+  Location loc = base.getLoc();
+  auto rhsType = cast<RankedTensorType>(base->getOperand(1).getType());
+  auto outType = cast<RankedTensorType>(base->getResult(0).getType());
+  std::optional<linalg::GenericOp> baseTruncOp;
+  if (fusionGroup.truncateOps) {
+    baseTruncOp = fusionGroup.truncateOps->front();
+  }
+
+  SmallVector<ReassociationIndices> reassoc;
+  reassoc.push_back({0, 1});
+  for (int i = 0, e = rhsType.getRank() - 1; i < e; ++i) {
+    reassoc.push_back({i + 2});
+  }
+
+  SmallVector<int64_t> rhsNewShape(rhsType.getShape());
+  rhsNewShape.insert(rhsNewShape.begin(), 1);
+  RankedTensorType concatRhsType =
+      RankedTensorType::get(rhsNewShape, rhsType.getElementType());
+
+  SmallVector<Value> rhsVals;
+  for (auto op : fusionGroup.contractionOps) {
+    Value thisRhs = op.getDpsInputOperand(1)->get();
+    Value expanded = rewriter.create<tensor::ExpandShapeOp>(loc, concatRhsType,
+                                                            thisRhs, reassoc);
+    rhsVals.push_back(expanded);
+  }
+
+  Value newRhs = rewriter.create<tensor::ConcatOp>(loc, 0, rhsVals);
+
+  SmallVector<int64_t> newShape(outType.getShape());
+  newShape.insert(newShape.begin(), rhsVals.size());
+  RankedTensorType concatOutType =
+      RankedTensorType::get(newShape, outType.getElementType());
+
+  Value baseOut = base.getDpsInitOperand(0)->get();
+  auto origFill = baseOut.getDefiningOp<linalg::FillOp>();
+  auto origEmpty =
+      origFill.getDpsInitOperand(0)->get().getDefiningOp<tensor::EmptyOp>();
+
+  auto newEmpty = rewriter.create<tensor::EmptyOp>(loc, concatOutType,
+                                                   origEmpty.getDynamicSizes());
+  auto newFill = rewriter.create<linalg::FillOp>(
+      loc, concatOutType, origFill.getDpsInputOperand(0)->get(),
+      newEmpty.getResult());
+
+  Value lhs = base->getOperand(0);
+
+  SmallVector<AffineMap> indexingMaps = base.getIndexingMapsArray();
+  SmallVector<utils::IteratorType> iteratorTypes = base.getIteratorTypesArray();
+  iteratorTypes.insert(iteratorTypes.begin(), utils::IteratorType::parallel);
+
+  indexingMaps[0] = indexingMaps[0].shiftDims(1);
+  indexingMaps[1] = indexingMaps[1].shiftDims(1).insertResult(
+      rewriter.getAffineDimExpr(0), 0);
+  indexingMaps[2] = indexingMaps[2].shiftDims(1).insertResult(
+      rewriter.getAffineDimExpr(0), 0);
+
+  linalg::GenericOp newGenericOp = rewriter.create<linalg::GenericOp>(
+      loc, concatOutType, ValueRange{lhs, newRhs}, newFill.getResult(0),
+      indexingMaps, iteratorTypes);
+  rewriter.cloneRegionBefore(base->getRegion(0), newGenericOp.getRegion(),
+                             newGenericOp.getRegion().begin());
+
+  Value fusedResult = newGenericOp.getResult(0);
+  if (fusionGroup.truncateOps) {
+    // Insert truncate operator.
+    auto truncType =
+        baseTruncOp.value()->getResult(0).getType().cast<RankedTensorType>();
```

Comment:
Can this be `cast<RankedTensorType>(...)`? The `x.cast<XXX>()` style is being deprecated.

https://mlir.llvm.org/deprecation/

> When casting attributes or type, use the free functions variants, e.g., dyn_cast<T>(x), isa<T>(x), etc. Use of the cast methods variants (e.g., x.dyn_cast<T>()) should be avoided in new code as we’ll remove these methods in the future.